flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/cleanup/CheckpointResourcesCleanupRunner.java

```diff
@@ -73,24 +76,24 @@
     // we have to have two separate futures because closeAsync relies on the completion of
     // getResultFuture which is always already completed but the cleanupFuture is only
     // instantiated when calling start
-    private CompletableFuture<Void> cleanupFuture;
+    @Nullable private CompletableFuture<Void> cleanupFuture;
```

Comment:
This is necessary to synchronize the `start` and `closeAsync` methods of the `CheckpointResourcesCleanupRunner`. `closeAsync` would be possible to complete before calling `start` which is problematic because the `start` method is called outside of registering the `CheckpointResourcesCleanupRunner` in the `Dispatcher`'s `JobManagerRunnerRegistry` which gets cleaned after the `CheckpointResourcesCleanupRunner.getResultFuture` is finished (which is the case right away when initializing the runner).

Explaining that, I realize that it's probably a cleaner approach where the resultfuture completes after the `start` method is called. That would enable us to merge the two future `closeFuture` and `cleanupFuture`. I will try that...